planetiler-core/src/main/java/com/onthegomap/planetiler/mbtiles/Compare.java

```diff
@@ -0,0 +1,160 @@
+ *
+ */
+public class Compare {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(Compare.class);
+
+  public static void main(String[] args) throws Exception {
+
+
+    Arguments arguments = Arguments.fromArgs(args);
+    String dbPath0 = arguments.getString("bench_mbtiles0", "the first mbtiles file", null);
+    String dbPath1 = arguments.getString("bench_mbtiles1", "the second mbtiles file", null);
+    boolean failOnFeatureDiff = arguments.getBoolean("bench_fail_on_feature_diff", "fail on feature diff", false);
+
+    try (
+      var db0 = Mbtiles.newReadOnlyDatabase(Path.of(dbPath0));
+      var db1 = Mbtiles.newReadOnlyDatabase(Path.of(dbPath1))
+    ) {
+      long tilesCount0 = getTilesCount(db0);
+      long tilesCount1 = getTilesCount(db1);
+      if (tilesCount0 != tilesCount1) {
+        throw new IllegalArgumentException(
+          "expected tiles count to be equal but tilesCount0=%d tilesCount1=%d".formatted(tilesCount0, tilesCount1)
+        );
+      }
+
+      int lastPercentage = -1;
+      long processedTileCounter = 0;
+      long tilesWithDiffs = 0;
+      try (var statement = db0.connection().prepareStatement("select tile_column, tile_row, zoom_level from tiles")) {
+        var rs = statement.executeQuery();
+        while (rs.next()) {
+          processedTileCounter++;
+          int x = rs.getInt("tile_column");
+          int y = rs.getInt("tile_row");
+          int z = rs.getInt("zoom_level");
+          TileCoord coord = TileCoord.ofXYZ(x, (1 << z) - 1 - y, z);
+
+          int currentPercentage = (int) (processedTileCounter * 100 / tilesCount0);
+          if (lastPercentage != currentPercentage) {
+            LOGGER.info("processed {}%", currentPercentage);
+          }
+          lastPercentage = currentPercentage;
+
+          var features0 = decode(gunzip(db0.getTile(coord)))
+            .stream()
+            .map(VectorTileFeatureForCmp::fromActualFeature)
+            .collect(Collectors.toSet());
+          var features1 = decode(gunzip(db1.getTile(coord)))
+            .stream()
+            .map(VectorTileFeatureForCmp::fromActualFeature)
+            .collect(Collectors.toSet());
+
+          if (!features0.equals(features1)) {
+            ++tilesWithDiffs;
+            boolean featureCountMatches = features0.size() == features1.size();
+            var msg = """
+              <<<
+              feature diff on coord %s - featureCountMatches: %b (%d vs %d)
+
+              additional in db0
+              ---
+              %s
+
+              additional in db1
+              ---
+              %s
+              >>>
+              """.formatted(
+              coord, featureCountMatches, features0.size(), features1.size(),
+              getDiffJoined(features0, features1, "\n"),
+              getDiffJoined(features1, features0, "\n"));
+
+            if (failOnFeatureDiff) {
+              throw new RuntimeException(msg);
```

Comment:
to fix sonar's complaint here and below, this could be a more specific exception like assertion error?